### ISSUE TITLE
Enums are not converted from snake_case

### ISSUE THREAD

**Issue by Zamiell** [timostamm/protobuf-ts#292]
Based on my brief testing, it does not seem that enum keys are converted from snake_case.

For example, I would expect the following proto code:

```proto
enum Status {
  launched = 1;
  init_game = 2;
  in_game = 3;
  in_replay = 4;
  ended = 5;
  quit = 6;
  unknown = 99;
}
```

To be converted to:

```ts
enum Status {
  Launched = 1,
  InitGame = 2,
  InGame = 3,
  InReplay = 4,
  Ended = 5,
  Quit = 6,
  Unknown = 99,
}
```

But instead, the enum keys are unchanged, and remain snake_case.

**Comment by timostamm** [comment]
True, we don't change enum value case - haven't considered that TBH. However:

Following the [style guide](https://developers.google.com/protocol-buffers/docs/style#enums), you should use:

```proto
enum Status {
  LAUNCHED = 1;
  INIT_GAME = 2;
  // ...
}
```

Which generates as:

```proto
enum Status {
  LAUNCHED = 1,
  INIT_GAME = 2,
  // ...
}
```

And I think that's not terrible. If you scroll down a bit in the TypeScript handbook, you'll notice an example with uppercase enum values as well.


By the way, all protobuf enums should really have a `UNSPECIFIED = 0` value in proto3 syntax, and all values should be prefixed by the enum name to prevent collisions. I recommend a linter like [buf](https://docs.buf.build/best-practices/style-guide#enums). So your enum should really be:


```proto
enum Status {
  STATUS_UNSPECIFIED = 0;
  STATUS_LAUNCHED = 1;
  STATUS_INIT_GAME = 2;
  // ...
}
```

(protobuf-ts drops the shared enum value prefix, so you aren't bothered by it in JS)

**Comment by Zamiell** [comment]
Tim, thanks for the quick reply.

> And I think that's not terrible. If you scroll down a bit in the TypeScript handbook, you'll notice an example with uppercase enum values as well.

That's true! However, the `LogLevel` enum is the only exception on the page - all the other enums are PascalCase.

<br>

## Most TypeScript Code uses PascalCase over SHOUTING_SNAKE_CASE

For TypeScript, I think that [Basarat's style guide](https://basarat.gitbook.io/typescript/styleguide#enum) is pretty popular, and it says to use PascalCase for enum members. He gives the following reason:

> Reason: Convention followed by TypeScript team i.e. the language creators e.g SyntaxKind.StringLiteral. Also helps with translation (code generation) of other languages into TypeScript.

This is confirmed by looking at code in the [TypeScript repository](https://github.com/microsoft/TypeScript/blob/main/src/server/project.ts#L3-L9). We can also see PascalCase enums when we look at code in the [Angular repository](https://github.com/angular/angular/blob/master/packages/compiler/src/expression_parser/parser.ts#L33-L49), or the [VSCode repository](https://github.com/microsoft/vscode/blob/main/src/vscode-dts/vscode.d.ts#L509-L525), or [BabylonJS](https://github.com/BabylonJS/Babylon.js/blob/0a87db0ef4ffc9aae1151d716dbf88f52ceb09f8/packages/dev/core/src/Materials/Node/Enums/nodeMaterialSystemValues.ts), to name a few big ones.

<br>

## PascalCase has functional advantages over SHOUTING_SNAKE_CASE

To be fair, I did notice some SHOUTING_SNAKE_CASE enums in these repos, but they are much less frequent. I think that this is for historical reasons.

In JavaScript, enum members have historically been SHOUTING_SNAKE_CASE for the same reason that it is conventional to write constants in SHOUTING_SNAKE_CASE - it represents non-immutability, communicating to the reader: "don't modify this variable!"

However, now that we have TypeScript, immutability no longer needs to be communicated via the variable name. We can just hover over the variable to find out that it is immutable / a constant. And the compiler will immediately notify us if we attempt to modify/mutate it.

In other words, it communicates more information to the reader when  you say that "this variable is a *type*" rather than "this variable is a *constant*".

<br>

## lower_snake_case

Additionally, consider the case where the authors of the protobuf file **do not** follow the Google protobuf style guide that you linked above. This happens to be the case for me: I'm using [the protobuf files for StarCraft 2](https://github.com/Blizzard/s2client-proto/tree/master/s2clientprotocol). These proto files use lower_snake_case enums, which results in `protobuf-ts` generating the following TypeScript code:

```ts
export enum Status {
    UNSPECIFIED$ = 0,
    launched = 1,
    init_game = 2,
    in_game = 3,
    in_replay = 4,
    ended = 5,
    quit = 6,
    unknown = 99
}
```

I can't influence the style in which the source proto files are written. I just need to use them as a base to try and generate idiomatic TypeScript code.

Maybe we could agree that:

- `PascalCase` enum members are idiomatic TypeScript.
- `SHOUTING_SNAKE_CASE` enum members are slightly non-idiomatic TypeScript.
- `lower_snake_case` enum members are **really** non-idiomatic TypeScript.

Having `protobuf-ts` make an enum with `lower_snake_case` is confusing / off-putting for TypeScript end-users. They might think to themselves: "Why is protobuf-ts converting everything to camelCase except for my enums! It must be bugged!"

I think that the tool should automatically convert both `lower_snake_case` and `SHOUTING_SNAKE_CASE` to `PascalCase`. But if it absolutely needs to leave alone `SHOUTING_SNAKE_CASE` for whatever reason, I think it should definitely at least convert `snake_case` to `PascalCase`.

<br>

> By the way, all protobuf enums should really have a UNSPECIFIED = 0 value in proto3 syntax

No worries - that was just a quick example that I manually typed out to showcase the casing.

**Comment by timostamm** [comment]
Thanks for your thoughts, @Zamiell. 

I _do_ agree that PascalCase fits well, and consistent naming rules are very nice to have. However, we need the original names for the JSON format - currently, we simply use them as is (and restore the shared prefix we strip). For message field names, we have a technique that allows arbitrary names for the generated properties, while still retaining small code size.

But we don't have that for enumerations. The implementation takes code size pretty seriously, but if people are interested in consistent PascalCase, and we find a good technique to implement that without taking up much code size, this could be a nice change for a major release.

**Comment by Zamiell** [comment]
The implementation should be trivial, right? I can submit a PR if you'll accept it.
